server/chat-plugins/lottery.js

```diff
@@ -0,0 +1,248 @@
+	}
+	destroy() {
+		delete lotteries[this.roomid];
+		this.write();
+	}
+	/**
+     * @param {User} user
+     */
+	addUser(user) {
+		if (!lotteries[this.roomid].participants[user.latestIp]) {
+			lotteries[this.roomid].participants[user.latestIp] = user.name;
+			this.write();
+			return true;
+		}
+		return false;
+	}
+	/**
+     * @param {User} user
+     */
+	removeUser(user) {
+		if (lotteries[this.roomid].participants[user.latestIp]) {
+			delete lotteries[this.roomid].participants[user.latestIp];
+			this.write();
+			return true;
+		}
+		return false;
+	}
+	getWinner() {
+		const winners = [];
+		const lotteryData = lotteries[this.roomid];
+		let participantsKeys = Object.keys(lotteryData.participants);
+		for (let i = 0; i < lotteryData.maxWinners; i++) {
+			const randomKey = participantsKeys[participantsKeys.length * Math.random() << 0];
+			const winner = lotteryData.participants[randomKey];
+			winners.push(winner);
+			participantsKeys = participantsKeys.filter(username => username !== winner);
+		}
+		return winners;
+	}
+}
+
+/** @type {ChatCommands} */
+exports.commands = {
+	lottery: {
+		'!': true,
+		''(target, room) {
+			if (!room) {
+				return this.errorReply("This can only be used in rooms.");
+			}
+			const lotteryData = lotteries[room.id];
+			if (!lotteryData) {
+				return this.errorReply("This room doesn't have a lottery running.");
+			}
+			return this.parse(`/join view-lottery-${room.id}`);
+		},
+		'!create': true,
+		create(target, room, user) {
+			if (!this.can('declare', null, room)) return;
+			if (!room) {
+				return this.errorReply("This can only be used in rooms.");
+			}
+			if (lotteries[room.id]) {
+				return this.errorReply(`There is already a lottery running.`);
+			}
+			const lottery = new Lottery(room.id);
+			let [maxWinners, name, ...markup] = target.split(',').map(val => val.trim());
+			if (!(maxWinners && name && markup.length)) {
+				return this.errorReply("You're missing a command parameter - see /help roomevents for this command's syntax.");
+			}
+			const maxWinnersNum = parseInt(maxWinners);
+			if (isNaN(maxWinnersNum)) {
+				return this.errorReply(`${maxWinners} is not a valid number.`);
+			}
+			if (maxWinnersNum < 1) {
+				return this.errorReply('The maximum winners should be at least 1.');
+			}
+			lottery.create(maxWinnersNum, name, markup.join(','));
+			this.sendReply('The lottery was successfully created.');
+			this.add(Chat.html`|raw|<div class="broadcast-blue"><b>${user.name} created the "${name}" lottery!</b></div>`);
+			this.modlog('LOTTERY CREATE');
+		},
+		'!delete': true,
+		delete(target, room, user) {
+			if (!this.can('declare', null, room)) return;
+			if (!room) {
+				return this.errorReply('This can only be used in rooms.');
+			}
+			const lotteryData = lotteries[room.id];
+			if (!lotteryData) {
+				return this.errorReply('This room does not have a lottery running.');
+			}
+			const lottery = new Lottery(room.id);
```

Comment:
I really fail to see why you're making a class for Lottery if you're not actually storing any information in it. You're keeping the state in a javascript object, and this class just provides a few methods without keeping any state. I would either keep track of a single Lottery object you use to store all the data in, or get rid of the OOP entirely.

Reply:
Refactored to low-level functions to manipulate the lottery.